### ISSUE TITLE
Magisk can't handle newest boot.img files for Android 13

### ISSUE THREAD

**Issue by joshudson** [topjohnwu/Magisk#6285]
Hours were lost because debug build was hidden behind an image. You may think this is plain but it isn't clear that image is a link at all.

Device: Pixel 6a
Android version: bluejay-t1b1.220819.007
Magisk version name: 38ab6858 (25203) (34) (D)
Magisk version code: 38ab6858 (25203) (D)

Magisk.log is useless:

- Device platform: arm64-v8a
- Installing: 38ab6858 (25203)
- Copying image to cache
Parsing boot image: [/data/user_de/0/com.topjohnwu.magisk/install/boot.img]
- Unpacking boot image
! Unsupported/Unknown image format
! Installation failed

Boot.img: https://github.com/topjohnwu/Magisk/files/9630380/boot.zip

The developer image https://developer.android.com/about/versions/12/dev-support-images doesn't work either.
[dev12.zip](https://github.com/topjohnwu/Magisk/files/9632809/dev12.zip)

**Comment by pndwal** [comment]
Why is that called 'bootloader-xxxx.img?... Is this bootloader or boot image??

**Comment by pndwal** [comment]
Man! ... Looks like your own linked doc ('developer...' link) says it's a 'bootloader update'!!! What are you doing???

**Comment by joshudson** [comment]
Trying to root my new phone. Due to a bug, the android 13 image has to be used to unlock bootloader, and the bootloader doesn't move backwards.

Unpacking the zip file yields s boot image and a system image.

**Comment by pndwal** [comment]
So what is the image you posted?... Seems named for bootloader, not boot!... Can you see boot.img, vendor _boot.img, etc?...

**Comment by joshudson** [comment]
@pndwal

````
joshua@nova:/tmp$ unzip ~/imagefiles/android/pixel6a/bluejay-t1b1.220819.007-factory-7f8d0622.zip    
Archive:  /home/joshua/imagefiles/android/pixel6a/bluejay-t1b1.220819.007-factory-7f8d0622.zip
   creating: bluejay-t1b1.220819.007/
  inflating: bluejay-t1b1.220819.007/radio-bluejay-g5123b-105339-220815-b-8947148.img  
  inflating: bluejay-t1b1.220819.007/flash-base.sh  
 extracting: bluejay-t1b1.220819.007/image-bluejay-t1b1.220819.007.zip  
  inflating: bluejay-t1b1.220819.007/flash-all.sh  
  inflating: bluejay-t1b1.220819.007/flash-all.bat  
  inflating: bluejay-t1b1.220819.007/bootloader-bluejay-bluejay-1.2-8946231.img  
````

And the contents of flash-all.sh:

````
joshua@nova:/tmp$ cat bluejay-t1b1.220819.007/flash-all.sh 
#!/bin/sh
# (snip apache 2.0 license for brevity)
if ! [ $($(which fastboot) --version | grep "version" | cut -c18-23 | sed 's/\.//g' ) -ge 3301 ]; then
  echo "fastboot too old; please download the latest version at https://developer.android.com/studio/releases/platform-tools.html"
  exit 1
fi
fastboot flash bootloader bootloader-bluejay-bluejay-1.2-8946231.img
fastboot reboot-bootloader
sleep 5
fastboot flash radio radio-bluejay-g5123b-105339-220815-b-8947148.img
fastboot reboot-bootloader
sleep 5
fastboot -w update image-bluejay-t1b1.220819.007.zip
````

File names are arbitrary anyway because the most convenient way to transfer to phone is `adb shell cat \> /sdcard/Downloads/boot.img < boot.img`

**Comment by pndwal** [comment]
I can see it's updating bootloader and baseband... Can't see boot... bootloader is NOT boot... 😝

**Comment by joshudson** [comment]
@pndwal : I agree with yuck face. I now see the actual problem; the firmware images provided by Google the obvious way are not complete images.

**Comment by melkor0** [comment]
@joshudson can you unzip the `image-bluejay-t1b1.220819.007.zip` file as well? In the current factory image (September 2022) is a `boot.img` file.

**Comment by joshudson** [comment]
@melkor0 : Yeah I did eventually figure that out. The instructions are completely bonkers. It appears the flash command is writing the zip file to the system partition, which is not what it actually does.

**Comment by pndwal** [comment]
What instructions?... The img files are in the needed format and flashed to partitions specified, eg per your example above:
> fastboot flash bootloader bootloader-bluejay-bluejay-1.2-8946231.img

/bootloader is the partition here; this won't touch /system ... Unless you use:
> fastboot flash system bootloader-bluejay-bluejay-1.2-8946231.img

... DON'T do that!

**Comment by joshudson** [comment]
@pndwal : `fastboot update image-bluejay-t1b1.220819.007.zip`

**Comment by pndwal** [comment]
But how does this 'appear that the flash command is writing the zip file to the system partition' only?

Stock updates usually include kernel fixes and enhancements needed to support system changes... and the kernel resides in boot.img...

**Comment by joshudson** [comment]
@pndwal : Understanding the exact details of how it went bad are rarely useful.

The last time I had to flash firmware there was no double-nested zip file and there was only a couple of files in the firmware image. Well, more than two but less than in that zip file. So I completely misread the thing as the system image is now a zip file, and the bootloader loads part of it into ram, ...

On that hypothesis, unpacking the zip file and replacing boot.img inside it makes no sense. Boot-packed archives are usually in a very specific format and won't unpack correctly if built with normal tools. Hint: if you unpack initrd.gz on a normal linux system with the normal way of unpacking .cpio.gz files, change a file in the resulting directory structure, and pack it back with the normal packing command sequence that generates one, even if you pass `-H newc` to match to cpio format used by the kernel it won't work because the kernel unpacker is picky and wants the archive inside out as compared to the normal correct ordering of a cpio archive.

Anyway I did in fact try to copy the inner zip file and feed magisk that; which didn't help of course.
